applications/FluidDynamicsApplication/custom_utilities/mass_conservation_utility.cpp

```diff
@@ -30,131 +30,195 @@ namespace Kratos
-        shift_for_correction = - water_volume_error / inter_area;
-        ShiftDistanceField( shift_for_correction );
-    }
+    KRATOS_INFO_IF("MassConservationUtility", mEchoLevel > 0) << "Volume error: " << mWaterVolumeError << std::endl;
 
     // assembly of the log message
     std::string output_line_timestep =  std::to_string(current_time) + "\t\t";
     output_line_timestep +=             std::to_string(neg_vol) + "\t\t";
     output_line_timestep +=             std::to_string(pos_vol) + "\t\t";
-    output_line_timestep +=             std::to_string(water_volume_error) + "\t\t";
+    output_line_timestep +=             std::to_string(mWaterVolumeError) + "\t\t";
     output_line_timestep +=             std::to_string(net_inflow_inlet) + "\t\t";
     output_line_timestep +=             std::to_string(net_inflow_outlet) + "\t\t";
-    output_line_timestep +=             std::to_string(inter_area) + "\t\t";
-    output_line_timestep +=             std::to_string(shift_for_correction) +"\n";
+    output_line_timestep +=             std::to_string(inter_area) + "\n";
+
     return output_line_timestep;
 }
 
 
 
-void MassConservationCheckProcess::ComputeVolumesAndInterface( double& positiveVolume, double& negativeVolume, double& interfaceArea ){
+double MassConservationUtility::ComputeDtForConvection(){
+
+    // a small step is set to avoid numerical problems
+    double time_step_for_convection = 1.0e-7;
+    const auto& r_comm = mrModelPart.GetCommunicator();
+
+    if ( mWaterVolumeError > 0.0 ){
+        // case: water volume was lost by mistake
+        mAddWater = true;
+        double water_outflow_over_boundary = OrthogonalFlowIntoAir( 1.0 );
+        if ( r_comm.GetDataCommunicator().SumAll( water_outflow_over_boundary ) ){
+            // checking if flow is sufficient (avoid division by 0)
+            if ( water_outflow_over_boundary > 1.0e-12 ){
+                time_step_for_convection = mWaterVolumeError / water_outflow_over_boundary;
+            }
+        } else {
+            KRATOS_ERROR << "Communication failed in MassConservationUtility::ComputeDtForConvection()";
+        }
+    }
+    else if ( mWaterVolumeError < 0.0 && mCorrectBackwards) {
+        // case: water volume was gained by mistake
+        mAddWater = false;
+        double water_inflow_over_boundary = OrthogonalFlowIntoAir( -1.0 );
+        if ( r_comm.GetDataCommunicator().SumAll( water_inflow_over_boundary ) ){
+            // checking if flow is sufficient (avoid division by 0)
+            if ( water_inflow_over_boundary > 1.0e-12 ){
+                time_step_for_convection = - mWaterVolumeError / water_inflow_over_boundary;
+            }
+        } else {
+            KRATOS_ERROR << "Communication failed in MassConservationUtility::ComputeDtForConvection()";
+        }
+    }
+    else {
+        // case: Exactly the correct volume of water is present
+        mAddWater = true;
```

Comment:
Why `mAddWater` is true if the the mass check is correct? `mAddWater` shouldn't be a `bool` as it might take three values. Instead we can use an `enum class` type or `int` (I think the first option eases the code reading).

Reply:
used an enum